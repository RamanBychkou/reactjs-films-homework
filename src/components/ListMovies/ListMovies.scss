.container {
  max-width: 82rem;
  margin: 0 auto;
}

.typeViewLines {
  flex-flow: column nowrap;
}

.moviesSelector {
  display: flex;
  align-items: baseline;
  justify-content: space-between;
}

.typeView button {
  border: none;
  color: #0aaee4;
  background: none;
  font-weight: bolder;
  font-size: 1.5rem;
  outline: none;
}

.moviesWrapper,
.moviesSelector ul {
  display: flex;
  flex-flow: row wrap;

  li {
    padding-bottom: 0.6rem;
    outline: none;

    &.active {
      border-bottom: 1px solid #0aaee4;
    }

    a button {
      color: #0aaee4;
      outline: none;
      border: none;
      cursor: pointer;
      background: none;
    }
  }
}

.movieList {
  background: #f2f2f2;
  padding: 5rem 1.5rem 1.5rem;
}

.moviesWrapper {
  justify-content: space-between;
}

.moviesSelector ul {
  justify-content: flex-start;

  li {
    margin-right: 1rem;
  }

  li,
  select {
    color: #0aaee4;
  }

  select,
  option {
    cursor: pointer;
  }

  select {
    background: #f2f2f2;
    border: none;
    width: 3.7rem;
    outline: none;

    option {
      border: none;
    }
  }
}

@media screen and (max-width: 42.5rem) {
  .moviesWrapper,
  .moviesSelector,
  .moviesSelector ul {
    justify-content: center;
  }
}

@media screen and (max-width: 31.25rem) {
  .movieList {
    padding-top: 7rem;
  }

  .typeView {
    display: none;
  }
}
