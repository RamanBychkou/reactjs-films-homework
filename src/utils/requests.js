import itemsIsLoadingAction from '../modules/isLoading/isLoadingAction';
import { itemsFetchDataSuccess, fetchVideoSuccess, clearResults } from '../modules/root/rootAction';
import setMainMovieDetails from '../modules/mainMovie/mainMovieAction';
import itemsHasErrored from '../modules/Error/errorAction';

const KEY = '75331f1a740385460b25b56203149aa8';

const requestsFilms = {
  fetchListMovies: (dispatch, query) => {
    let url;
    // const id = query.replace(/\/[genre]/g);
    switch (query) {
      case 'coming_soon':
        url = `https://api.themoviedb.org/3/movie/upcoming?api_key=${KEY}&language=en-US&page=1`;
        break;
      case 'trending':
        url = `https://api.themoviedb.org/3/movie/popular?api_key=${KEY}&language=en-US&page=1`;
        break;
      case '/':
        url = `https://api.themoviedb.org/3/movie/popular?api_key=${KEY}&language=en-US&page=1`;
        break;
      case 'top_rated':
        url = `https://api.themoviedb.org/3/movie/top_rated?api_key=${KEY}&language=en-US&page=1`;
        break;
      default:
        url = `https://api.themoviedb.org/3/discover/movie?api_key=${KEY}&with_genres=${query}`;
    }
    dispatch(itemsIsLoadingAction(true));
    return fetch(url)
      .then(resp => resp.json())
      .then((data) => {
        dispatch(itemsIsLoadingAction(false));
        if (data.ok === false) {
          throw new Error('Something wrong');
        }
        return dispatch(
          itemsFetchDataSuccess({
            page: data.page,
            results: data.results,
          }),
        );
      })
      .catch(err => dispatch(itemsHasErrored(err)));
  },
  getMainMovieDetails: (dispatch, id) => {
    dispatch(itemsIsLoadingAction(true));
    return fetch(
      `https://api.themoviedb.org/3/movie/${id}?api_key=75331f1a740385460b25b56203149aa8&language=en-US`,
    )
      .then(resp => resp.json())
      .then((data) => {
        dispatch(itemsIsLoadingAction(false));
        if (data.ok === false) {
          throw new Error('Something wrong');
        }
        return dispatch(setMainMovieDetails(data));
      })
      .catch(err => dispatch(itemsHasErrored(err)));
  },

  fetchVideo: (dispatch, id) => {
    dispatch(itemsIsLoadingAction(true));
    return fetch(`https://api.themoviedb.org/3/movie/${id}/videos?api_key=${KEY}&language=en-US`)
      .then(resp => resp.json())
      .then((data) => {
        dispatch(itemsIsLoadingAction(false));
        if (data.ok === false) {
          throw new Error('Something wrong');
        }
        let videoId = '';
        data.results.some((curr) => {
          if (curr.type === 'Trailer') {
            videoId = curr.key;
            return true;
          }
          return false;
        });
        return dispatch(fetchVideoSuccess(videoId));
      })
      .catch(err => dispatch(itemsHasErrored(err)));
  },

  fetchSearchResults: (dispatch, query) => {
    dispatch(itemsIsLoadingAction(true));
    return fetch(
      `https://api.themoviedb.org/3/search/movie?api_key=${KEY}&language=en-US&query=${query.replace(
        '%20',
        '',
      )}&page=1&include_adult=false`,
    )
      .then(resp => resp.json())
      .then((data) => {
        if (data.results.length === 0) {
          dispatch(clearResults());
          throw new Error('Nothing found');
        }
        dispatch(itemsIsLoadingAction(false));
        return dispatch(
          itemsFetchDataSuccess({
            page: data.page,
            results: data.results,
          }),
        );
      })
      .catch(err => dispatch(itemsHasErrored(err.message)));
  },
};

export default requestsFilms;
