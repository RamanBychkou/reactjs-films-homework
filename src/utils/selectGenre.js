const selectGenre = function selectGenre(props, e) {
  const {
    genres, fetchListMovies, error, clearError,
  } = props;
  const genre = e.target.value || e.target.textContent.replace(/[^A-Za-z\s]/g, '');
  let genreId;
  genres.some((a) => {
    if (genre === a.name) {
      genreId = a.id;
      return true;
    }
    return false;
  });
  if (error !== undefined && error !== '') {
    clearError(false);
  }
  return fetchListMovies(genreId);
};

export const keydonwGenres = (props, e) => {
  if (e.key === 'Enter') {
    selectGenre(props, e);
  }
  return props;
};

export default selectGenre;
