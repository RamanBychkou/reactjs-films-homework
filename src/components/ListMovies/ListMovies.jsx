import React from 'react';
import shortid from 'shortid';
import selectGenre from '../../utils/selectGenre';

import HoverMovieCard from '../HoverMovieCard/HoverMovieCard';

import * as stylerate from '../Star/Star.scss';
import style from './ListMovies.scss';


const changeMainFilm = (props, e) => {
  e.preventDefault();
  // Change mainFilmOverwie height
  document.querySelector('#mainFilmOverwie').style.maxHeight = '10rem';

  const name = e.currentTarget.textContent;
  const { changeMainMovie, results } = props;
  let newMainFilm;
  results.some((curr) => {
    if (curr.title === name) {
      newMainFilm = curr;
      return true;
    }
    return false;
  });
  return changeMainMovie(newMainFilm);
};

const ListMovies = (props) => {
  const { results, genres } = props;
  if (results.length > 0) {
    return results.map((curr) => {
      const imageLink = `https://image.tmdb.org/t/p/w300/${curr.poster_path}`;
      const imgAlt = curr.title;
      const genresRow = curr.genre_ids.map((c) => {
        let currentGenre;
        genres.some((a) => {
          if (c === a.id) { currentGenre = a.name; return true; }
          return false;
        });
        return currentGenre;
      }).join(', 1').split('1').map(cur => (
        <li key={shortid.generate()}>
          <span role="button" onClick={selectGenre.bind({ props })}>{ cur}</span>
        </li>
      ));

      return (
        <article
          className={style.movieContainer}
          key={shortid.generate()}
        >
          <figure>
            <div>
              <img src={imageLink} alt={imgAlt} />
            </div>
            <figcaption>
              <div className={style.movieTitle}>
                <a href="#" id="movieTitle" onClick={changeMainFilm.bind(null, props)}>{curr.title}</a>
                <span className={stylerate.rate}>{curr.vote_average}</span>
              </div>
              <div className={style.movieGenres}>
                <ul>
                  {genresRow}
                </ul>
              </div>
            </figcaption>
          </figure>
          <HoverMovieCard
            overview={curr.overview}
            genres={genresRow}
            title={curr.title}
            rate={curr.vote_average}
          />
        </article>
      );
    });
  }
  return <h2>Ooops</h2>;
};

export default ListMovies;

export { changeMainFilm };
