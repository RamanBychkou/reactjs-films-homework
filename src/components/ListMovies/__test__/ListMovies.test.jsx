import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import fetchMock from 'fetch-mock';
import ListMovies from '../index';
import { keydonwGenres } from '../../../utils/selectGenre';
import requestsFilms from '../../../utils/requests';


const genres = [
  { id: 28, name: 'Action' },
  { id: 12, name: 'Adventure' },
  { id: 16, name: 'Animation' },
  { id: 35, name: 'Comedy' },
  { id: 80, name: 'Crime' },
  { id: 99, name: 'Documentary' },
  { id: 18, name: 'Drama' },
  { id: 10751, name: 'Family' },
  { id: 14, name: 'Fantasy' },
  { id: 36, name: 'History' },
  { id: 27, name: 'Horror' },
  { id: 10402, name: 'Music' },
  { id: 9648, name: 'Mystery' },
  { id: 10749, name: 'Romance' },
  { id: 878, name: 'Science Fiction' },
  { id: 10770, name: 'TV Movie' },
  { id: 53, name: 'Thriller' },
  { id: 10752, name: 'War' },
  { id: 37, name: 'Western' },
];

test('ListMovies renders correctly with fill Array results', () => {
  const results = [{
    adult: false,
    backdrop_path: '/7RyHsO4yDXtBv1zUU3mTpHeQ0d5.jpg',
    genre_ids: [12, 878, 28],
    id: 299534,
    original_language: 'en',
    original_title: 'Avengers: Endgame',
    overview: "After the devastating events of Avengers: Infinity War, the universe is in ruins due to the efforts of the Mad Titan, Thanos. With the help of remaining allies, the Avengers must assemble once more in order to undo Thanos' actions and restore order to the universe once and for all, no matter what consequences may be in store.",
    popularity: 323.106,
    poster_path: '/or06FN3Dka5tukK1e9sl16pB3iy.jpg',
    release_date: '2019-04-24',
    title: 'Avengers: Endgame',
    video: false,
    vote_average: 8.6,
    vote_count: 4484,
  }];


  const renderer = new ShallowRenderer();
  renderer.render(<ListMovies results={results} genres={genres} />);
  expect(renderer.getRenderOutput()).toMatchSnapshot();
});

test('ListMovies renders correctly with empty Array results', () => {
  const results = [];
  const renderer = new ShallowRenderer();
  const result = renderer.render(<ListMovies results={results} genres={genres} />);
  expect(result).toMatchSnapshot();
});

describe('ListMovies: selectGenre', () => {
  afterEach(() => {
    fetchMock.reset();
    fetchMock.restore();
  });

  it('ListMovies: kewDown on genre, expect props', async () => {
    const idGenre = 35;
    fetchMock
      .getOnce(`https://api.themoviedb.org/3/discover/movie?api_key=75331f1a740385460b25b56203149aa8&with_genres=${idGenre}`, {
        headers: { 'content-type': 'application/json' },
        body: { page: 1, results: [1, 2, 3], status: 'ok' },
      }).catch(err => err);
    const props = {
      genres: [{ id: 31, name: 'Action' }, { id: 35, name: 'Drama' }],
      fetchMoviesOnGenre: requestsFilms.fetchMoviesOnGenre,
    };
    const e = {
      key: 'a',
      target: {
        id: 'test',
        textContent: 'Drama',
      },
    };
    const result = keydonwGenres(props, e);
    expect(result).toEqual(props);
  });

  it('ListMovies: kewDown Enter, expect data', async () => {
    const idGenre = 35;
    fetchMock
      .getOnce(`https://api.themoviedb.org/3/discover/movie?api_key=75331f1a740385460b25b56203149aa8&with_genres=${idGenre}`, {
        headers: { 'content-type': 'application/json' },
        body: { page: 1, results: [1, 2, 3], status: 'ok' },
      }).catch(err => err);
    const props = {
      genres: [{ id: 31, name: 'Action' }, { id: 35, name: 'Drama' }],
      fetchMoviesOnGenre: requestsFilms.fetchMoviesOnGenre,
    };
    const e = {
      key: 'Enter',
      target: {
        id: 'test',
        textContent: 'Drama',
      },
    };
    const result = keydonwGenres(props, e);
    expect(result).toEqual(props);
  });

  it('ListMovies: kewDown on genre', async () => {
    const idGenre = 35;
    fetchMock
      .getOnce(`https://api.themoviedb.org/3/discover/movie?api_key=75331f1a740385460b25b56203149aa8&with_genres=${idGenre}`, {
        headers: { 'content-type': 'application/json' },
        body: { page: 1, results: [1, 2, 3], status: 'ok' },
      }).catch(err => err);
    const props = {
      genres: [{ id: 31, name: 'Action' }, { id: 35, name: 'Drama' }],
      fetchMoviesOnGenre: requestsFilms.fetchMoviesOnGenre,
    };
    const e = {
      key: 'a',
      target: {
        id: 'test',
        textContent: 'Drama',
      },
    };
    const result = keydonwGenres(props, e);
    expect(result).toEqual(props);
  });
});
