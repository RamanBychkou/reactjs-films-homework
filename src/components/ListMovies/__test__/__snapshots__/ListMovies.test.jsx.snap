// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListMovies renders correctly with empty Array results 1`] = `
<h2>
  Ooops
</h2>
`;

exports[`ListMovies renders correctly with fill Array results 1`] = `
Array [
  <article
    className="movieContainer"
    id={299534}
  >
    <figure>
      <div>
        <img
          alt="Avengers: Endgame"
          src="https://api.ballotpedia.org/v3/thumbnail/"
        />
      </div>
      <figcaption>
        <div
          className="movieTitle"
        >
          <button
            href="#"
            id="movieTitle"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex="0"
            type="button"
          >
            Avengers: Endgame
          </button>
          <span>
            8.6
          </span>
        </div>
        <div
          className="movieGenres"
        >
          <ul>
            <li>
              <button
                onClick={[Function]}
                onKeyDown={[Function]}
                type="button"
              >
                Adventure, 
              </button>
            </li>
            <li>
              <button
                onClick={[Function]}
                onKeyDown={[Function]}
                type="button"
              >
                Science Fiction, 
              </button>
            </li>
            <li>
              <button
                onClick={[Function]}
                onKeyDown={[Function]}
                type="button"
              >
                Action
              </button>
            </li>
          </ul>
        </div>
      </figcaption>
    </figure>
    <hoverMovieCard
      fetchVideo={[Function]}
      genres={
        Array [
          <li>
            <button
              onClick={[Function]}
              onKeyDown={[Function]}
              type="button"
            >
              Adventure, 
            </button>
          </li>,
          <li>
            <button
              onClick={[Function]}
              onKeyDown={[Function]}
              type="button"
            >
              Science Fiction, 
            </button>
          </li>,
          <li>
            <button
              onClick={[Function]}
              onKeyDown={[Function]}
              type="button"
            >
              Action
            </button>
          </li>,
        ]
      }
      movies={Object {}}
      overview="After the devastating events of Avengers: Infinity War, the universe is in ruins due to the efforts of the Mad Titan, Thanos. With the help of remaining allies, the Avengers must assemble once more in order to undo Thanos' actions and restore order to the universe once and for all, no matter what consequences may be in store."
      rate={8.6}
      title="Avengers: Endgame"
    />
  </article>,
]
`;
